### ISSUE TITLE
Split repo in multiple components ?

### ISSUE THREAD

**Issue by hadim** [twosigma/beakerx#5135]
Would that make sense to have each component in a different repo ?

For example `kernel/base` would allow a strict separation of the kernel from the code that evaluates languages. That would make easier packaging (in conda-forge for example).

Each language could also live in their own repo. We do it for the `scijava/scripting-*` packages.

**Comment by hadim** [comment]
Beaker kernel and Beaker Jupyter extensions could also be separated from each other.

**Comment by kno10** [comment]
I would appreciate if the JVM parts do not depend on conda and nodejs, and individual kernels can be built separately.

Because I don't use conda, but apt-get for managing python; and I'd rather not have another package manager with npm in the loop...

I'm mostly interested in seeing a good Java or Groovy kernel for jupyter, to use with ELKI.

I don't care much about fancy JavaScript UI enhancements (that won't work on my data anyway).
Last I checked out Beaker, I ended up not using it because of dependency hell. And even the beakerx re-write still seems to pull in a lot of unnecessary dependencies. kernel base depends on "h2database" and "fluent-hc", but I don't see either being used anywhere; and the clojure kernel (which I will never use) pulls in dead projects like sonatype/eclipse "aether"...
I would also appreciate if more dependencies (e.g. tablesaw, which is 30 MB and includes MANY other libraries, including Apache math3, fastutil, ... all without proper dependency and version management - this will be a version pain for you, eventually) were made optional.
If I disable everything but the groovy and java kernel, change `--sys-prefix` to `--user` it appears that I can install the kernels, and Groovy has working tab completion. Unfortunately, it doesn't appear to be very stable yet.

**Comment by hadim** [comment]
That would also allow having different Maven artifacts : one for the base kernel and others for each language. So people are free to pick whatever they want without having to ship unused code.

**Comment by hadim** [comment]
For now, I generate a base kernel artifact with the following code : 

```bash
git clone https://github.com/twosigma/beakerx.git
cd beakerx/kernel/base/
gradle publishToMavenLocal
```

**Comment by altavir** [comment]
It is a bad idea. Many different repos are hard to maintain. And you do not need this. Gradle allows to publish separate artifacts without splitting repository.
you can use `gradle :kernel:base:<whatever>` instead of `cd`. If the artifacts would be deployed to `mavenCentral` or any other repository, then you could just download it with maven or gradle without cloning the repository.

**Comment by hadim** [comment]
What do you mean by hard to maintain ? Because I think it's the opposite actually.

Having small multiple repositories is easier for the maintainers because the issue and PRs are well separated and each member of the dev team can so better focus on one part of the project. 

Releases can be decoupled and so it's faster to quickly iterate and make a new release with new features on one part of the project without touching another component. Continuous integration, such as Travis, is also easier to setup and debug since builds are faster.

For all the reasons above I think it also makes things easier to understand for developers who want to implement a new kernel using Beakerx (such as me).

This is the way the ImageJ/Scijava is organized and it's a very good example of the "divide to conquer" philosophy : one repo for one component. Check the GitHub organization to give you an idea of the usefulness of this : https://github.com/imagej and https://github.com/scijava

I think @ctrueden, the lead developer of ImageJ/Scijava, will agree with me. But I might be missing some strong arguments.

Of course, it's possible to have multiple maven artifacts even using one repository but I think the profit of having multiple repositories goes well beyond that : it allows maintainers and contributors to work better together as a community.

**Comment by kno10** [comment]
I don't think it is generally necessary to split a repository (although the beakerx repository would benefit from being recreated without the history of the old codebase).
If you have strong version interdependencies, then it may be better to keep things in the same repository to keep everything in sync. Then you don't need to commit to two repositories, and update two repositories in lockstep...

My main concern is the **overuse of dependencies** in `kernel/base` itself. In particular central components should only use minimal dependencies. That makes it harder to reuse the base kernel component for other kernels, if it pulls in massive jars like tablesaw that include (without dependency management) other large jars. Dependencies need to be kept to a minimum, and if possible should be made optional.

In this case, I can imagine it would be good to make a repository with the "core" kernel parts only. Say base, java and maybe groovy; then maintain additional kernels in separate repositories (e.g. a C or C++ kernel will likely not share anything with this codebase; a clojure kernel may have a small userbase, and could simply depend on the base kernel artifact once it is stable), maintain UI extensions (beakerx/beakerx with python and javascript) separately; JVM plotting separate; and maybe also make the whole data translation layer yet another repository (some users may not need this)?

I'd rather not have a SciJava dependency from the base either; I don't see the need to have a plugin mechanism at that level, and many features of SciJava will not work here (no AWT UI etc.); and for my own use cases, SciJava appears to be too limited. Some parts of scijava don't seem to make offer any functinoality yet at all (scijava 3d). SciJava scripting doesn't appear to support tab-completion, but without tab-completion the kernels will likely not make users very happy anymore... (but of course, there could be a scijava-scripting-kernel) and I'd rather keep the number of dependencies small.

**Comment by hadim** [comment]
I agree with you on the overuse of dependencies.

About Scijava, I never said a Scijava dep should be added to Beakerx. If the community wants to, I would be happy to help of course but I am perfectly fine with Beakerx being totally independent of Scijava.

That being said Scijava is extremely modular (one repo = one component) and so it's extremely easy to pick only whatever you want and need for your project. So, for example, there is no need to use scijava AWT related code (it lives in a separate component). About tab-completion, it's definitely a feature we want to add in order to have autocompletion to work in the upcoming [Scijava Jupyter Kernel](https://github.com/hadim/scijava-jupyter-kernel).

**Comment by scottdraves** [comment]
>  the JVM parts do not depend on conda and nodejs

it does not really depend on conda afaik, that is just a simplification for the documentation.

however, nodejs is required to build the JS and some of the features of the JVM plugins depend on that, such as setting the imports and classpath from the notebook metadata, our widgets, and autotranslation.  we will generally consider these an integral part of BeakerX.

>  overuse of dependencies in kernel/base

there are some extras here, that's just an oversight and will be fixed, thank you for bringing it up.

>  it may be better to keep things in the same repository to keep everything in sync

indeed there is a cost to multiple repos and i would currently be inclined to keep one repo but publish multiple jars/packages from it, so that there are reusable components of minimal size and minimize version skew problems. longer term as APIs become stable, splitting becomes more beneficial.  

> individual kernels can be built separately.

i believe this is the case.

**Comment by hadim** [comment]
I agree with that plan @scottdraves .

Did you consider releasing artifacts soon ? That would make easier for me to setup CI with Travis and for people to build my kernel.

I think it's fine to make a release early even if you tag it as alpha or beta. Your base kernel works well enough.

**Comment by ctrueden** [comment]
> I think @ctrueden, the lead developer of ImageJ/Scijava, will agree with me. But I might be missing some strong arguments.

Every way has its pros and cons.

Personally, I am a fan of individual fine-grained components managed by a bill of materials to keep the versions straight. And separate Git repositories. But there are certainly advantages to keeping everything in one repository. In the interest of time, I'll just say that in general, I agree with @scottdraves that for early dev it's often helpful to have a single repo, and as things stabilize it is often useful to split things up into separate repos with independent versions.
